### DISCUSSION TITLE
Is there easier way to get failed Response from fetch request with AbortSignal?

### DISCUSSION THREAD

**Discussion by fifzteen** [visgl/deck.gl#6859] [Q&A] [answered]
Hi!
I uses TileLayer, and fetch tiles from remote. 
I want to get the failed `Response` object if requests failed. (`TileLayer.onTileError`  passes `Error` object.)
So I try override `props.loadOptions.fetch` to my function

```
import { fetchFile } from '@loadres.gl/core

new TileLayer({
  loadOptions: {
    fetch: async (url, options) => {
      try {
        const response = await fetchFile(url, options)
        if (!response.ok) throw response
        return response 
      } catch (error) {
        myCuntomHandling(error)
        throw error
      }
    },
  },
  ...
})
```

This doesn't work.
I found that `TileLayer.getTileData` passes AbortSignal to `Layer.props.fetch`, and `Layer.props.fetch` force overrides loadOptions to Object when signal exists.

```
// a part of TileLayer.getTileData
const {data, getTileData, fetch} = this.props;
const {signal} = tile;

...

if (tile.url) {
  return fetch(tile.url, {propName: 'data', layer: this, signal}); // passes abortSignal to fetch
}

// a part of Layer.defaultProps.fetch
if (signal) {
  loadOptions = {
    ...loadOptions,
    fetch: {
      ...loadOptions?.fetch, // fetch is overrided to Object!
      signal
    }
  };
}
```

To get failed Response, Do I need to pass custom `props.fetch` to TileLayer?
(Layer.props.fetch seems a bit complicated, so I would not want to change that if I could...)
Is there easier way to get failed Response from fetch request with AbortSignal?

**Comment by fifzteen** [comment, marked as answer]
I solved this problem. (in my usage)

```js
import { load, fetchFile } from '@loaders.gl/core'

new TileLayer({
  ...
  fetch: (url, { layer, loaders, loadOptions, signal }) => {
    loaders = loaders || layer.props.loaders
    loadOptions = loadOptions || layer.getLoadOptions()
    
    // directly wrote my fetch function
    loadOptions.fetch = async (url, options) => {
      const response = await fetchFile(url, {
        ...options,
        signal,
        // options originally set to "loadOptions.fetch"
        method: 'GET',
        mode: 'cors',
        credentials: 'include',
      })
      if (!response.ok) throw resopnse
      return response
    }

    return load(url, loaders, loadOptions)
  },
  onTileError: (error) => {
    myCuntomHandling(error)
  }
})
```
* I cannot find a way to set my fetch function to `loadOptions.fetch` and additionally pass AbortSignal to `loadOptions.fetch` in `props.fetch`, so write my fetch function directly in `props.fetch` and stay `loadOptions.fetch` default.
* I need to set some params to `loadOptions.fetch` (CORS and so on). Therefore `resourceManager` in original `props.fetch` is always ignored. So I deleted all codes about `resourceManager` from `props.fetch`.

**Reply by Pessimistress** [reply]
TileLayer never uses `resourceManager` - it's basically a cache mechanism for resource sharing between layers, while TileLayer manages its own cache.

You could just skip `loadOptions.fetch` and implement it like this:

```js
import {parse} from '@loaders.gl/core';

const fetchFunc = async (url, { layer, loaders, loadOptions, signal }) => {
    const response = await fetch(url, {
        signal,
        method: 'GET',
        mode: 'cors',
        credentials: 'include',
    })
    if (!response.ok) {
        // handle errors
    }

    return parse(response, loaders, loadOptions)
};

new TileLayer({
    fetch: fetchFunc,
    // ...other props
});
```

Cc @ibgreen this is a case where `loaderOptions.fetch` cannot replace `props.fetch`.

**Reply by fifzteen** [reply]
My memo: When MVTLayer, it has special loader `MVTWorkerLoader` on its props, so need a line below.

```
loaders = loaders || layer.props.loaders
```
